frame/state-trie-migration/src/lib.rs

```diff
@@ -0,0 +1,1514 @@
+		///
+		/// This can only be valid if it is generated from the local node, which means only
+		/// validators can generate this call.
+		///
+		/// The `item_limit` is the maximum number of items that can be read whilst ensuring that
+		/// the migration does not go over `Self::unsigned_limits().size`.
+		///
+		/// The `witness_size` should always be equal to `Self::unsigned_limits().size` and is only
+		/// used for weighing.
+		#[pallet::weight(Pallet::<T>::dynamic_weight(*item_limit, *witness_size))]
+		pub fn continue_migrate_unsigned(
+			origin: OriginFor<T>,
+			item_limit: u32,
+			witness_size: u32,
+			_witness_task: MigrationTask<T>,
+		) -> DispatchResultWithPostInfo {
+			ensure_none(origin)?;
+			let chain_limits =
+				Self::unsigned_limits().ok_or("unsigned limit not set, tx not allowed.")?;
+			ensure!(witness_size == chain_limits.size, "wrong witness data");
+
+			let mut task = Self::migration_process();
+			// pre-dispatch and validate-unsigned already assure this.
+			debug_assert_eq!(task, _witness_task);
+
+			// we run the task with the given item limit, and the chain size limit..
+			task.migrate_until_exhaustion(MigrationLimits {
+				size: chain_limits.size,
+				item: item_limit,
+			});
+
+			// .. and we assert that the size limit must have been fully met with the given item
+			// limit.
+			assert!(task.fully_complies_with(chain_limits));
+
+			Self::deposit_event(Event::<T>::Migrated(
+				task.dyn_top_items,
+				task.dyn_child_items,
+				MigrationCompute::Unsigned,
+			));
+			MigrationProcess::<T>::put(task);
+
+			Ok(().into())
+		}
+
+		/// Continue the migration for the given `limits`.
+		///
+		/// The dispatch origin of this call can be any signed account.
+		///
+		/// This transaction has NO MONETARY INCENTIVES. calling it will only incur transaction fees
+		/// on the caller, with no rewards paid out.
+		///
+		/// The sum of the byte length of all the data read must be provided for up-front
+		/// fee-payment and weighing.
+		#[pallet::weight(
+			// the migration process
+			Pallet::<T>::dynamic_weight(limits.item, * real_size)
+			// rest of the operations, like deposit etc.
+			+ T::WeightInfo::continue_migrate()
+		)]
+		pub fn continue_migrate(
+			origin: OriginFor<T>,
+			limits: MigrationLimits,
+			real_size: u32,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			let max_limits = T::SignedMigrationMaxLimits::get();
+			ensure!(
+				limits.size <= max_limits.size && limits.item <= max_limits.item,
+				"max signed limits not respected"
+			);
+
+			// ensure they can pay more than the fee.
+			let deposit = T::SignedDepositPerItem::get().saturating_mul(limits.item.into());
+			ensure!(T::Currency::can_slash(&who, deposit), "not enough funds");
+
+			let mut task = Self::migration_process();
+			task.migrate_until_exhaustion(limits);
+
+			// ensure that the migration witness data was correct.
+			if real_size != task.dyn_size {
```

Comment:
I think this constraint is too strong, when calculating the real_size we do it offchain, and depending on the initialisation and the previous extrinsics, the touched items can be different resulting in a different size.

I think caller should call with an overestimate of real_size otherwise it is quite difficult to call this. But indeed in the current code because we don't recalculate the accurate weight and just do `Pays::No` it can result in an attack. To prevent this I think we should return the accurate weight in the post_info. so that the unused weight is removed from the block weight.

Reply:
Sure, will leave this open to do this later. 

Note that I initially didn't do that since even if you overestimate, then you have to decide by how much you want to over-estimate, which cannot be clearly answered. I think it is fine for the bot to fail in rare cases. At the end of the day, it is probably being funded by the network, and it can deal with it.

Reply:
finally fixed this now.